lib/digital_signal/digital_signal.c

```diff
@@ -80,94 +160,385 @@ uint32_t digital_signal_get_edge(DigitalSignal* signal, uint32_t edge_num) {
+    furi_assert(signal);
+
+    if(!signal->edge_cnt) {
+        return;
+    }
+
+    /* Configure gpio as output */
+    signal->internals->gpio = gpio;
+    furi_hal_gpio_init(
+        signal->internals->gpio, GpioModeOutputPushPull, GpioPullNo, GpioSpeedVeryHigh);
+
+    /* single signal, add a temporary, terminating edge at the end */
+    signal->edge_timings[signal->edge_cnt++] = 10;
+    digital_signal_prepare_arr(signal);
+
+    digital_signal_setup_dma(signal);
+    digital_signal_setup_timer();
+    digital_signal_start_timer();
+
+    while(!LL_DMA_IsActiveFlag_TC2(DMA1)) {
+    }
+
+    digital_signal_stop_timer();
+    digital_signal_stop_dma();
+
+    signal->edge_cnt--;
+}
+
+void digital_sequence_alloc_signals(DigitalSequence* sequence, uint32_t size) {
+    sequence->signals_size = size;
+    sequence->signals = malloc(sequence->signals_size * sizeof(DigitalSignal*));
+    sequence->signals_prolonged = malloc(sequence->signals_size * sizeof(bool));
+}
+
+void digital_sequence_alloc_sequence(DigitalSequence* sequence, uint32_t size) {
+    sequence->sequence_used = 0;
+    sequence->sequence_size = size;
+    sequence->sequence = malloc(sequence->sequence_size);
+    sequence->send_time = 0;
+}
+
+DigitalSequence* digital_sequence_alloc(uint32_t size, const GpioPin* gpio) {
+    DigitalSequence* sequence = malloc(sizeof(DigitalSequence));
+
+    sequence->gpio = gpio;
+    sequence->bake = false;
+
+    digital_sequence_alloc_signals(sequence, 32);
+    digital_sequence_alloc_sequence(sequence, size);
+
+    return sequence;
+}
+
+void digital_sequence_free(DigitalSequence* sequence) {
+    furi_assert(sequence);
+
+    if(!sequence) {
+        return;
+    }
+
+    free(sequence->signals);
+    free(sequence->sequence);
+    free(sequence);
+}
+
+void digital_sequence_set_signal(
+    DigitalSequence* sequence,
+    uint8_t signal_index,
+    DigitalSignal* signal) {
+    furi_assert(sequence);
+    furi_assert(signal);
+    furi_assert(signal_index < sequence->signals_size);
 
-    while(!LL_DMA_IsActiveFlag_TC2(DMA1))
-        ;
+    sequence->signals[signal_index] = signal;
+    signal->internals->gpio = sequence->gpio;
+    signal->internals->reload_reg_remainder = 0;
+
+    digital_signal_prepare_arr(signal);
+}
+
+void digital_sequence_set_sendtime(DigitalSequence* sequence, uint32_t send_time) {
+    sequence->send_time = send_time;
+}
+
+void digital_sequence_add(DigitalSequence* sequence, uint8_t signal_index) {
+    furi_assert(sequence);
+    furi_assert(signal_index < sequence->signals_size);
+
+    if(sequence->sequence_used >= sequence->sequence_size) {
+        sequence->sequence_size += 256;
+        sequence->sequence = realloc(sequence->sequence, sequence->sequence_size);
+    }
+
+    sequence->sequence[sequence->sequence_used++] = signal_index;
+}
+
+#if defined(DIGITAL_SIGNAL_PORTABLE_CODE)
+
+static void digital_signal_update_dma(DigitalSignal* signal) {
```

Comment:
i am so sorry, you are absolutely right. 
unfortunately i pushed an incomplete C code. didnt notice because i flashed an old version...

here the disassembly of the two variants next to each other.
![grafik](https://user-images.githubusercontent.com/479231/210056826-f0ac9493-d510-4791-aa28-54b5cf66da9b.png)

it clearly shows the larger footprint you noticed, but also the lower re-enable latency if the DMA finished its transfer.

Reply:
okay optimized a bit more
![grafik](https://user-images.githubusercontent.com/479231/210070489-54cdc815-1b24-43e9-88b9-6c0154f344c3.png)

instruction count difference is 13 instructions, however i am not happy with the stack usage. could be optimized.

Reply:
okay further optimized the code.
![grafik](https://user-images.githubusercontent.com/479231/210086851-14f5c804-7a93-4279-b6ef-cfc4736ed91f.png)
i am skipping DMA channel 1 initialization. will have to double-check that there are no side effects.
now the C implementation (right) might be in the usable range of delay. did not check with NfcA emulation to prove this is working out.